src/vs/editor/contrib/stickyScroll/browser/stickyScrollModelProvider.ts

```diff
@@ -0,0 +1,314 @@
+
+				const status = await modelProvider.computeModel(
+					textModel,
+					textModelVersionId,
+					token
+				);
+
+				switch (status) {
+					case Status.CANCELED:
+						this._store.clear();
+						return null;
+					case Status.VALID:
+						return modelProvider.stickyModel;
+				}
+			}
+
+			return null;
+		});
+	}
+}
+
+interface IStickyModelCandidateProvider {
+	get stickyModel(): StickyModel | null;
+	get providerModelPromise(): CancelablePromise<any> | null;
+	set stopWatch(stopWatch: StopWatch | null);
+	set updateDebounceInfo(updateDebounceInfo: IFeatureDebounceInformation | null);
+	set updateScheduler(updateScheduler: Delayer<StickyModel | null> | null);
+
+	/**
+	 * Method which computes the sticky model and returns a status to signal whether the sticky model has been successfully found
+	 * @param textmodel text-model of the editor
+	 * @param modelVersionId version ID of the text-model
+	 * @param token cancellation token
+	 * @returns a promise of a status indicating whether the sticky model has been successfully found
+	 */
+	computeModel(textmodel: ITextModel, modelVersionId: number, token: CancellationToken): Promise<Status> | Status;
+}
+
+abstract class StickyModelCandidateProvider implements IStickyModelCandidateProvider {
+
+	private _providerModelPromise: CancelablePromise<any> | null = null;
+	protected _stickyModel: StickyModel | null = null;
+	private _updateScheduler: Delayer<StickyModel | null> | null = null;
+	private _updateDebounceInfo: IFeatureDebounceInformation | null = null;
+	private _stopWatch: StopWatch | null = null;
+
+	constructor() { }
+
+	get stickyModel(): StickyModel | null {
+		return this._stickyModel;
+	}
+
+	get providerModelPromise(): CancelablePromise<any> | null {
+		return this._providerModelPromise;
+	}
+
+	set updateScheduler(updateScheduler: Delayer<StickyModel | null> | null) {
+		this._updateScheduler = updateScheduler;
+	}
+
+	set updateDebounceInfo(updateDebounceInfo: IFeatureDebounceInformation | null) {
+		this._updateDebounceInfo = updateDebounceInfo;
+	}
+
+	set stopWatch(stopWatch: StopWatch | null) {
+		this._stopWatch = stopWatch;
+	}
+
+	private _invalid(): Status {
+		this._stickyModel = null;
+		return Status.INVALID;
+	}
+
+	public computeModel(textModel: ITextModel, modelVersionId: number, token: CancellationToken): Promise<Status> | Status {
+		if (!this.isProviderValid(textModel)) {
+			return this._invalid();
+		}
+		const providerModelPromise = this._providerModelPromise = createCancelablePromise(token => this.createModelFromProvider(textModel, modelVersionId, token));
+
+		return providerModelPromise.then(providerModel => {
+			if (!this.isModelValid(providerModel)) {
+				return this._invalid();
+
+			} else if (providerModelPromise === this._providerModelPromise && this._updateDebounceInfo && this._stopWatch && this._updateScheduler) {
+				const newValue = this._updateDebounceInfo.update(textModel, this._stopWatch.elapsed());
```

Comment:
I think it would also be ok to do `this._updateDebounceInfo.update(textModel, this._stopWatch.elapsed());` after `return this.createStickyModel`

Then it can go outside of `computeModel` (into the code that calls it) and there's no need to have `stopWatch` and `updateDebounceInfo` on `IStickyModelCandidateProvider`